File: server/src/main/java/org/dreamexposure/discal/server/network/dbotsgg/UpdateDBotsData.java
package org.dreamexposure.discal.server.network.dbotsgg;

import org.dreamexposure.discal.core.logger.LogFeed;
import org.dreamexposure.discal.core.logger.object.LogObject;
import org.dreamexposure.discal.core.object.BotSettings;
import org.dreamexposure.discal.core.utils.GlobalConst;
import org.dreamexposure.discal.server.DisCalServer;
import org.json.JSONObject;

import java.util.Timer;
import java.util.TimerTask;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class UpdateDBotsData {
    private static Timer timer;

    public static void init() {
        if ("true".equalsIgnoreCase(BotSettings.UPDATE_SITES.get())) {
            timer = new Timer(true);
            timer.schedule(new TimerTask() {
                @Override
                public void run() {
                    updateSiteData();
                }
            }, GlobalConst.oneHourMs);
        }
    }

    public static void shutdown() {
        if (timer != null) {
            timer.cancel();
        }
    }

    private static void updateSiteData() {
        try {
            final JSONObject json = new JSONObject()
                .put("guildCount", DisCalServer.getNetworkInfo().getTotalGuildCount())
                .put("shardCount", DisCalServer.getNetworkInfo().getExpectedClientCount());

            OkHttpClient client = new OkHttpClient();

            RequestBody body = RequestBody.create(GlobalConst.JSON, json.toString());
            Request request = new Request.Builder()
                .url("https://discord.bots.gg/api/v1/bots/265523588918935552/stats")
                .post(body)
                .header("Authorization", BotSettings.D_BOTS_GG_TOKEN.get())
                .header("Content-Type", "application/json")
                .build();

            Response response = client.newCall(request).execute();

            if (response.code() != GlobalConst.STATUS_SUCCESS)
                LogFeed.log(LogObject.forDebug("Failed to update DBots.gg stats", "Body: " +
                    (response.body() != null ? response.body().string() : "null")));
        } catch (Exception e) {
            LogFeed.log(LogObject.forException("Failed to update DBots.gg stats", e, UpdateDBotsData.class));
        }
    }
}

File: server/src/main/kotlin/org/dreamexposure/discal/server/network/dbotsgg/UpdateDBotsData.kt
package org.dreamexposure.discal.server.network.dbotsgg

import kotlinx.coroutines.reactor.awaitSingleOrNull
import kotlinx.coroutines.reactor.mono
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.dreamexposure.discal.core.config.Config
import org.dreamexposure.discal.core.logger.LOGGER
import org.dreamexposure.discal.core.utils.GlobalVal
import org.dreamexposure.discal.server.business.NetworkStatusService
import org.json.JSONObject
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.stereotype.Component
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.Duration

@Component
@ConditionalOnProperty("bot.integrations.update-bot-sites", havingValue = "true")
class UpdateDBotsData(
    private val networkStatusService: NetworkStatusService,
    private val httpClient: OkHttpClient
) : ApplicationRunner {
    private val token = Config.SECRET_INTEGRATION_D_BOTS_GG_TOKEN.getString()

    private fun update() = mono {
        val status = networkStatusService.getNetworkStatus()

        Mono.fromCallable {
            val json = JSONObject()
                .put("guildCount", status.totalGuildsCount)
                .put("shardCount", status.expectedShardCount)

            val body = json.toString().toRequestBody(GlobalVal.JSON)
            val request = Request.Builder()
                .url("https://discord.bots.gg/api/v1/bots/265523588918935552/stats")
                .post(body)
                .header("Authorization", token)
                .header("Content-Type", "application/json")
                .build()

            httpClient.newCall(request).execute()
        }.doOnNext { response ->
            if (response.code != GlobalVal.STATUS_SUCCESS) {
                LOGGER.debug("Failed to update DBots.gg stats | Body: ${response.body.string()}")
                response.body.close()
                response.close()
            }
        }.onErrorResume {
            Mono.empty()
        }.awaitSingleOrNull()
    }

    override fun run(args: ApplicationArguments?) {
        Flux.interval(Duration.ofHours(1))
            .flatMap { update() }
            .subscribe()
    }
}
